front/src/applications/operationalStudies/views/Scenario/components/MacroEditor/utils.ts

```
@@ -151,6 +152,100 @@ export const deleteMacroNodeByNgeId = async (
   }
 };
 
+const getNoteLabelTextsFromIds = (netzgrafikDto: NetzgrafikDto, labelIds: number[]): string[] =>
+  labelIds
+    .map((id) => netzgrafikDto.labels.find((l) => l.id === id)?.label)
+    .filter((l): l is string => Boolean(l));
+
+export const createMacroNote = async (
+  state: MacroEditorState,
+  dispatch: AppDispatch,
+  netzgrafikDto: NetzgrafikDto,
+  note: NoteIndexed
+) => {
+  try {
+    const response = await dispatch(
+      osrdEditoastApi.endpoints.postProjectsByProjectIdStudiesAndStudyIdScenariosScenarioIdMacroNotes.initiate(
+        {
+          projectId: state.projectId,
+          studyId: state.studyId,
+          scenarioId: state.scenarioId,
+          macroNoteBatchForm: {
+            macro_notes: [
+              {
+                x: Math.round(note.x),
+                y: Math.round(note.y),
+                title: note.title,
+                text: note.text,
+                labels: getNoteLabelTextsFromIds(netzgrafikDto, note.labelIds),
+              },
+            ],
+          },
+        }
+      )
+    ).unwrap();
+
+    const createdNote = response.macro_notes[0];
+    state.setDbIdForNote(note.ngeId, createdNote.id);
+  } catch (e) {
+    console.error(e);
+  }
+};
+
+export const updateMacroNote = async (
+  state: MacroEditorState,
+  dispatch: AppDispatch,
+  netzgrafikDto: NetzgrafikDto,
+  note: NoteIndexed
+) => {
+  try {
+    const dbId = state.getDbIdForNote(note.ngeId);
+    if (!dbId) return;
```

Same as the delete case: I think we should throw an error instead of silently doing nothing if we don't find a DB ID here.

---

Done